### PR TITLE
*: Rework reporting of invalid and wrong PeerIds

### PR DESCRIPTION
**Pull request by mxinden**
Replaces https://github.com/libp2p/rust-libp2p/pull/2428.

### CODE REVIEW

**Comment by rkuhn** [comment]
Thanks a lot for pushing this over the line, @mxinden!

File: core/src/connection/pool.rs [hunk 1]
```diff
@@ -777,49 +777,42 @@ where
...
                         })
                         // Check peer is not local peer.
                         .and_then(|()| {
-                            if self.local_id == peer_id {
-                                Err(Error::InvalidPeerId)
+                            if self.local_id == obtained_peer_id {
+                                Err(PendingConnectionError::WrongPeerId {
+                                    obtained: obtained_peer_id,
+                                    endpoint: endpoint.clone(),
+                                })
```
Comment (bkchr) [inline core/src/connection/pool.rs:815]: Why is that a `WrongPeerId`

**Review by bkchr — COMMENTED** [review]
In general I think this pr broke something. We have this check that prints a warning when we connect to a bootnode and that provides a different than expected peer id. Any idea what could be broken. See the following issue: https://github.com/paritytech/substrate/issues/11709

File: core/src/connection/pool.rs (hunk 1 above)
Comment (bkchr) [reply core/src/connection/pool.rs:815]: CC @mxinden

Comment (tomaka) [reply core/src/connection/pool.rs:815]: You're connecting to yourself, which isn't allowed. It's true that the error is unclear, but this was an `InvalidPeerId` error before this PR and in general has been an error for a very long time now, so I don't see how this specific change could have broken anything.

Comment (bkchr) [reply core/src/connection/pool.rs:815]: Yeah, I also looked over this changeset and didn't find anything that could cause the issue we are seeing. However, I think that after the corresponding libp2p update in Substrate this error started to appear. There is also some internal parity devops issue about the Kusama bootnodes reporting a wrong peerid, but devops assured to me multiple times that all keys are correct. So yeah, I'm assuming now that maybe something changed it libp2p :shrug:

**Pull request merged**